### PR TITLE
Forming surface bonds

### PR DESCRIPTION
**Pull request by davidfarinajr**
### Motivation or Problem
We recently discovered molecules in our models that have multiple bonds to the surface through a single adatom
Screen Shot 2021-07-02 at 10 14 31 AM
https://rmg.mit.edu/database/molecule/1%20O%20u0%20p2%20c0%20%7B2%2CS%7D%20%7B3%2CS%7D%0A2%20X%20u0%20p0%20c0%20%7B1%2CS%7D%0A3%20X%20u0%20p0%20c0%20%7B1%2CS%7D%0A
I don't think we should be making these molecules.  

 

### Description of Changes
added `is_bonded_to_surface` Atom and GroupAtom methods and modified the `_apply` ReactionRecipe function so that it raises an `InvalidActionError` if trying to form a surface bond to an atom already bonded to the surface

### Testing
Not tested yet

### Reviewer Tips
see if these molecules are still in models

### CODE REVIEW

File: rmgpy/molecule/molecule.pxd [hunk 1]
```diff
@@ -79,6 +79,8 @@ cdef class Atom(Vertex):
     cpdef bint is_nos(self)
 
     cpdef bint is_surface_site(self)
+
+    cpdef bind is_bonded_to_surface(self)
```
Comment (rwest) [inline rmgpy/molecule/molecule.pxd:83]: This PR is consistent with the rest of the file so I'm not requesting you change it before we merge, but do you have any idea why `molecule.pxd` doesn't have the `except -2`  declarations but `group.pxd` does?

**Review by rwest — APPROVED** [review]
This looks OK to me (just looking at the code, not running it) but did you figure out why it's needed? What reaction recipe was messing up, and can it be fixed?

File: rmgpy/molecule/molecule.pxd (hunk 1 above)
Comment (rwest) [reply rmgpy/molecule/molecule.pxd:83]:
```suggestion
    cpdef bint is_bonded_to_surface(self)
```

Comment (davidfarinajr) [reply rmgpy/molecule/molecule.pxd]: Thanks, I just fixed the type with force push, and nope, I am not sure why that is

**Review by mazeau — DISMISSED** [review]
I know the tests haven't officially passed yet, but this looks very good to me! I wonder what changed for this to be needed though....

**Comment by davidfarinajr** [comment]
I declared the GroupAtom `is_bonded_to_surface` function in the Group class and not the GroupAtom class. It's just one of those days 🤦 .

**Comment by davidfarinajr** [comment]
🤞 attempt number 3

**Comment by davidfarinajr** [comment]
> This looks OK to me (just looking at the code, not running it) but did you figure out why it's needed? What reaction recipe was messing up, and can it be fixed?

Looks like it was `Surface_Dissociation` which has this template
```
 *1--*2                 *1     *2
  |            ---->    ||      |
~*3~ + ~*4~~           ~*3~ + ~*4~~
```

It was applied to a bidentate molecule that had *2 bonded to `X`

**Comment by rwest** [comment]
Hmm. Thanks. Are we stuck without a way to declare absence of a bond? i.e. how do we put "*2 cannot be bonded to an X" into the top level template?

**Comment by rwest** [comment]
"Cannot convert rmgpy.molecule.group.GroupBond to rmgpy.molecule.group.GroupAtom"  Really not your day! 😢

**Comment by davidfarinajr** [comment]
lol, I told @mazeau that I would drop out if the test fail again so I guess this is good bye 👋

**Comment by davidfarinajr** [comment]
I'll fix it, get ready for another force push and attempt number 4

**Comment by davidfarinajr** [comment]
> Hmm. Thanks. Are we stuck without a way to declare absence of a bond? i.e. how do we put "*2 cannot be bonded to an X" into the top level template?

Yes.  The most elegant solution would be to somehow put that in the top level group so that those bidentate reactants don't match the template and we don't try to create product structures.  With this PR, that bidentate reactant will match and the family will try to generate product structures, but it will fail to do so due to the `InvalidActionError` when applying the template.  However, I still think this is a better and more generalizable solution then adding forbidden structs to each family which are only checked after the products are made and the reaction recipe is applied.

**Review by mazeau — APPROVED** [review]
this still looks good to me, and all the test pass! I say it's merge time

**Pull request merged**